sound/soc/sof/intel/cnl.c

```diff
@@ -72,6 +72,9 @@ irqreturn_t cnl_ipc4_irq_thread(int irq, void *context)
 				snd_sof_ipc_get_reply(sdev);
 				snd_sof_ipc_reply(sdev, data->primary);
 
+				/* Let DSP know that we have finished processing the message */
+				cnl_ipc_host_done(sdev);
```

Comment:
Interesting. But does this close the race window completely? What if a reply arrives, but the sending thread sends a new message from the host to the DSP before we ack here? Maybe even before the interrupt triggers?

Reply:
We can not have two tx messages at the same time. they are serialized with a higher level mutex, the next message can only proceed if the previous got the reply.
The problem is that we have multiple races in firmware as well and the two independent channels are made dependent of each other.

No, the checking of the tx channel that it is free is not going to fix this race as there is still (I have seen that with only the busy patch) that the tx channel from host was free, we sent a message, fw handled and tried to send a reply, but the kernel still have not managed to ack the reply. The fw delayed message handling is broken, so we ended up with IPC timeout.